File: facesdk_java_tests/src/test/java/SearchTests.java
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

import com.regula.facesdk.webclient.ApiException;
import com.regula.facesdk.webclient.gen.model.*;

import java.io.IOException;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

class SearchTests {
    private UUID groupId;
    private UUID personId;
    private static final String nameA = "Person A";
    Map<String, Object> baseMetadata = new HashMap<>() {{
        put("description", "This is a test group");
    }};

    private void standardAssertions(SearchResult response, UUID personId) {
        assertNotNull(response, "No response received");
        assertNotNull(response.getPersons(), "Person list is null");
        SearchPerson person = null;
        for (SearchPerson p : response.getPersons()) {
            if (Objects.requireNonNull(p.getId()).equals(personId)) {
                person = p;
                break;
            }
        }
        assertNotNull(person, "Person ID " + personId + " not found in the persons list");
        assertNotNull(Objects.requireNonNull(person.getImages()).get(0).getPath(), "The image path for person ID " + personId + " is empty");
    }

    @BeforeEach
    void setUp() throws ApiException, IOException {
        GroupToCreate groupToCreate = new GroupToCreate();
        String testGroupName = "test";
        groupToCreate.setName(testGroupName);
        groupToCreate.setMetadata(baseMetadata);
        Group createdGroup = PathsConfig.faceSdk.groupApi.createGroup(groupToCreate);
        groupId = createdGroup.getId();
        PersonFields personFields = new PersonFields();
        personFields.setName(nameA);
        personFields.setGroups(Collections.singletonList(groupId));
        personFields.setMetadata(baseMetadata);
        Person createdPerson = PathsConfig.faceSdk.personApi.createPerson(personFields);
        personId = createdPerson.getId();
        byte[] imageContent = PathsConfig.readImageBytes(PathsConfig.FACE3_PATH);
        AddImageToPersonRequestImage image = new AddImageToPersonRequestImage();
        image.setContent(imageContent);
        image.setContentType("image/jpeg");
        AddImageToPersonRequest imageFields = new AddImageToPersonRequest();
        imageFields.setImage(image);
        PathsConfig.faceSdk.personApi.addImageToPerson(personId, imageFields);
    }


    @AfterEach
    void tearDown() throws ApiException {
        // Clean up - delete the person and group
        if (personId != null) {
            PathsConfig.faceSdk.personApi.deletePerson(personId);
        }
        if (groupId != null) {
            PathsConfig.faceSdk.groupApi.deleteGroup(groupId);
        }
    }

    @Test
    void shouldSearchWithLimitAndThreshold() {
        try {
            SearchRequest searchRequest = new SearchRequest();
            searchRequest.setGroupIds(Arrays.asList(groupId));
            byte[] imageContent = PathsConfig.readImageBytes(PathsConfig.FACE1_PATH);
            AddImageToPersonRequestImage image = new AddImageToPersonRequestImage();
            image.setContentType("image/jpeg");
            image.setContent(imageContent);
            searchRequest.setImage(image);
            searchRequest.setLimit(10);
            searchRequest.setThreshold(0.8f);

            SearchResult response = PathsConfig.faceSdk.searchApi.search(searchRequest);
            standardAssertions(response, personId);
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
            throw e;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }


    @Test
    void shouldSearchWithBackgroundAndCrop() {
        try {
            SearchRequest searchRequest = new SearchRequest();
            searchRequest.setGroupIds(Arrays.asList(groupId));
            byte[] imageContent = PathsConfig.readImageBytes(PathsConfig.FACE1_PATH);
            AddImageToPersonRequestImage image = new AddImageToPersonRequestImage();
            image.setContentType("image/jpeg");
            image.setContent(imageContent);
            searchRequest.setImage(image);
            OutputImageParams oiParams = new OutputImageParams();
            oiParams.setBackgroundColor(Arrays.asList(128, 128, 128));
            Crop crop = new Crop();
            crop.setType(FaceImageQualityAlignType.fromValue(0));
            crop.setPadColor(Arrays.asList(0, 0, 0));
            crop.setSize(Arrays.asList(300, 400));
            oiParams.setCrop(crop);
            searchRequest.outputImageParams(oiParams);
            SearchResult response = PathsConfig.faceSdk.searchApi.search(searchRequest);
            standardAssertions(response, personId);
            SearchPerson person = response.getPersons().stream()
                    .filter(p -> p.getId().equals(personId))
                    .findFirst()
                    .orElse(null);
            assertTrue(person != null && person.getDetection() != null
                            && person.getDetection().getCrop() != null,
                    "Crop data not found in person's detection");
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
            throw e;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }


    @Test
    void shouldSearchWithResize() {
        try {
            SearchRequest searchRequest = new SearchRequest();
            searchRequest.setGroupIds(Arrays.asList(groupId));
            byte[] imageContent = PathsConfig.readImageBytes(PathsConfig.FACE1_PATH);
            AddImageToPersonRequestImage image = new AddImageToPersonRequestImage();
            image.setContentType("image/jpeg");
            image.setContent(imageContent);
            searchRequest.setImage(image);
            ResizeOptions resOpt = new ResizeOptions();
            resOpt.setHeight(867);
            resOpt.setWidth(1300);
            resOpt.setQuality(99);
            Objects.requireNonNull(searchRequest.getImage()).setResizeOptions(resOpt);
            SearchResult response = PathsConfig.faceSdk.searchApi.search(searchRequest);
            standardAssertions(response, personId);
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
            throw e;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }


    @Test
    void shouldSearchWithUrl() {
        try {
            SearchRequest searchRequest = new SearchRequest();
            searchRequest.setGroupIds(Arrays.asList(groupId));
            AddImageToPersonRequestImage image = new AddImageToPersonRequestImage();
            image.setImageUrl("https://img.freepik.com/free-photo/portrait-beautiful-blond-woman-with-trendy-hairstyle_23-2149430891.jpg?t=st=1712577121~exp=1712577721~hmac=e4fb2fa9517e9bf0953bcc7eda15059d131fec68541bee602b8885f6e99bbc9b");
            searchRequest.setImage(image);
            SearchResult response = PathsConfig.faceSdk.searchApi.search(searchRequest);

            assertNotNull(response, "No response received");
            assertTrue(Objects.requireNonNull(response.getPersons()).stream().anyMatch(person -> Objects.requireNonNull(person.getId()).equals(personId)), "Expected person not found in search results");
        } catch (ApiException e) {
            System.err.println("ApiException occurred: " + e.getResponseBody());
            throw e;
        }
    }
}

File: facesdk_cs_tests/facesdk_cs_tests/SearchTests.cs
namespace facesdk_cs_tests;

[TestFixture]
public class SearchTests
{
    private Guid _groupId;
    private Guid _personId;
    private const string NameA = "Person A";

    private readonly Dictionary<string, object> _baseMetadata = new Dictionary<string, object>
    {
        { "description", "This is a test group" }
    };

    private static void StandardAssertions(SearchResult response, Guid personId)
    {
        Assert.That(response, Is.Not.Null, "No response received");
        Assert.That(response.Persons, Is.Not.Null, "Person list is null");
        var person = response.Persons.FirstOrDefault(p => p.Id == personId);
        Debug.Assert(person != null, nameof(person) + " != null");
        Assert.That(person.Images[0].Path, Is.Not.Null, $"The image path for person ID {personId} is empty");
    }

    [SetUp]
    public void SetUp()
    {
        var groupToCreate = new GroupToCreate
        {
            Name = "test",
            Metadata = _baseMetadata
        };
        var createdGroup = PathsConfig.FaceSdk.GroupApi.CreateGroup(groupToCreate);
        _groupId = createdGroup.Id;

        var personFields = new PersonFields
        {
            Name = NameA,
            Groups = new List<Guid> { _groupId },
            Metadata = _baseMetadata
        };
        var createdPerson = PathsConfig.FaceSdk.PersonApi.CreatePerson(personFields);
        _personId = createdPerson.Id;

        var imageContent = PathsConfig.ReadImageBytes(PathsConfig.Face3Path);
        var image = new ImageFieldsImage
        {
            ContentType = "image/jpeg",
            Content = imageContent
        };
        var imageFields = new ImageFields { Image = image };
        PathsConfig.FaceSdk.PersonApi.AddImageToPerson(_personId, imageFields);
    }

    [TearDown]
    public void TearDown()
    {
        if (_personId != Guid.Empty)
            PathsConfig.FaceSdk.PersonApi.DeletePerson(_personId);
        if (_groupId != Guid.Empty)
            PathsConfig.FaceSdk.GroupApi.DeleteGroup(_groupId);
    }

    [Test]
    public void ShouldSearchWithLimitAndThreshold()
    {
        try
        {
            var searchRequest = new SearchRequest
            {
                GroupIds = new List<Guid> { _groupId },
                Image = new ImageFieldsImage
                {
                    ContentType = "image/jpeg",
                    Content = PathsConfig.ReadImageBytes(PathsConfig.Face1Path)
                },
                Limit = 10,
                Threshold = 0.8f
            };

            var response = PathsConfig.FaceSdk.SearchApi.Search(searchRequest);
            StandardAssertions(response, _personId);
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine($"ApiException occurred: {e.ErrorCode + e.Message}");
            throw;
        }
        catch (IOException e)
        {
            throw new InvalidOperationException(e.Message, e);
        }
    }

    [Test]
    public void ShouldSearchWithBackgroundAndCrop()
    {
        try
        {
            var searchRequest = new SearchRequest
            {
                GroupIds = new List<Guid> { _groupId },
                Image = new ImageFieldsImage
                {
                    ContentType = "image/jpeg",
                    Content = PathsConfig.ReadImageBytes(PathsConfig.Face1Path)
                },
                OutputImageParams = new OutputImageParams
                {
                    BackgroundColor = new List<int> { 128, 128, 128 },
                    Crop = new Crop
                    {
                        Type = FaceImageQualityAlignType.ALIGN_1x1, // Adjust as needed
                        PadColor = new List<int> { 0, 0, 0 },
                        Size = new List<int> { 300, 400 }
                    }
                }
            };

            var response = PathsConfig.FaceSdk.SearchApi.Search(searchRequest);
            StandardAssertions(response, _personId);
            var person = response.Persons.FirstOrDefault(p => p.Id == _personId);
            Assert.IsTrue(person is { Detection.Crop: not null },
                "Crop data not found in person's detection");
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine($"ApiException occurred: {e.ErrorCode + e.Message}");
            throw;
        }
        catch (IOException e)
        {
            throw new InvalidOperationException(e.Message, e);
        }
    }

    [Test]
    public void ShouldSearchWithResize()
    {
        try
        {
            var searchRequest = new SearchRequest
            {
                GroupIds = new List<Guid> { _groupId },
                Image = new ImageFieldsImage
                {
                    ContentType = "image/jpeg",
                    Content = PathsConfig.ReadImageBytes(PathsConfig.Face1Path),
                    ResizeOptions = new ResizeOptions
                    {
                        Height = 867,
                        Width = 1300,
                        Quality = 99
                    }
                }
            };

            var response = PathsConfig.FaceSdk.SearchApi.Search(searchRequest);
            List<Guid> personIds = response.Persons.Select(person => person.Id).ToList();
            foreach (var id in personIds)
            {
                TestContext.Out.WriteLine(id);
            }

            TestContext.Out.WriteLine(personIds.Count);
            StandardAssertions(response, _personId);
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine($"ApiException occurred: {e.ErrorCode + e.Message}");
            throw;
        }
        catch (IOException e)
        {
            throw new InvalidOperationException(e.Message, e);
        }
    }

    [Test]
    public void ShouldSearchWithUrl()
    {
        try
        {
            var searchRequest = new SearchRequest
            {
                GroupIds = new List<Guid> { _groupId },
                Image = new ImageFieldsImage
                {
                    ImageUrl = "https://img.freepik.com/free-photo/portrait-beautiful-blond-woman-with-trendy-hairstyle_23-2149430891.jpg?t=st=1712577121~exp=1712577721~hmac=e4fb2fa9517e9bf0953bcc7eda15059d131fec68541bee602b8885f6e99bbc9b"
                }
            };

            var response = PathsConfig.FaceSdk.SearchApi.Search(searchRequest);

            Assert.That(response, Is.Not.Null, "No response received");
            List<Guid> personIds = response.Persons.Select(person => person.Id).ToList();
            foreach (var id in personIds)
            {
                TestContext.Out.WriteLine(id);
            }

            TestContext.Out.WriteLine(personIds.Count);
        }
        catch (ApiException e)
        {
            Console.Error.WriteLine($"ApiException occurred: {e.ErrorCode + e.Message}");
            throw;
        }
    }
}